src/hotspot/share/prims/unsafe.cpp

```diff
@@ -783,21 +783,23 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
 
 UNSAFE_ENTRY(void, Unsafe_Unpark(JNIEnv *env, jobject unsafe, jobject jthread)) {
   if (jthread != nullptr) {
-    ThreadsListHandle tlh;
-    JavaThread* thr = nullptr;
-    oop java_thread = nullptr;
-    (void) tlh.cv_internal_thread_to_JavaThread(jthread, &thr, &java_thread);
-    if (java_thread != nullptr) {
-      // This is a valid oop.
-      if (thr != nullptr) {
-        // The JavaThread is alive.
-        Parker* p = thr->parker();
-        HOTSPOT_THREAD_UNPARK((uintptr_t) p);
-        p->unpark();
+    oop thread_oop = JNIHandles::resolve_non_null(jthread);
+    if (java_lang_Thread::thread(thread_oop) != nullptr) {
+      ThreadsListHandle tlh; // Provides memory barrier
+      if (java_lang_Thread::thread(thread_oop) != nullptr) {
+        JavaThread* thr = nullptr;
+        // We verified that java_lang_Thread::thread(thread_oop) != nullptr
+        // before and after creating tlh above so quick_mode can be used in
+        // this conversion call.
+        if (tlh.cv_internal_thread_to_JavaThread(jthread, &thr, nullptr, true /* quick_mode */)) {
```

Comment:
Sorry I'm having a mental blank here now. Why do we even need to call this if we can access `java_lang_Thread::thread(thread_oop)` after the TLH was created?

Reply:
As a caller of cv_internal_thread_to_JavaThread we do not know the implementation of quick_mode.

As the implementation is now, we still want the asserts.

Reply:
In particular we want this assert from `cv_internal_thread_to_JavaThread()`:
```
  assert(includes(java_thread), "must be");
```
because that verifies our assumptions from this call-site that `quick_mode` is okay.

Also, it is possible that a JavaThread that's racing to exit will make it past `ensure_join()`
after our check on L789 above and run into this block in the conversion function:
```
  JavaThread *java_thread = java_lang_Thread::thread(thread_oop);
  if (java_thread == nullptr) {
    // The java.lang.Thread does not contain a JavaThread* so it has not
    // run enough to be put on a ThreadsList or it has exited enough to
    // make it past ensure_join() where the JavaThread* is cleared.
    return false;
  }
```

Because we (the observer thread) captured the target JavaThread in the
ThreadsListHandle, it's safe for us to use the `JavaThread*`, but I've always
been a fan of bailing on operations like these as soon as we detect that the
`JavaThread*` is exiting.

Reply:
But ignoring the asserts `cv_internal_thread_to_JavaThread()` serves absolutely no purpose when we could request quick-mode - and we don't need to know what `cv_internal_thread_to_JavaThread` does - we know how thread startup and exit is managed and we know what a TLH guarantees . Sure we can bail sooner but there is no need to, so that is not required. In quick-mode, barring the assertion for includes, it all boils down to nothing! We don't need the Thread_oop passed out as we already have it before we start; we don't even need the JavaThread* passed out as we already have it too! So what does it actually do for us? Nothing. We could just have:
```
UNSAFE_ENTRY(void, Unsafe_Unpark(JNIEnv *env, jobject unsafe, jobject jthread)) {
  if (jthread != nullptr) {
    oop thread_oop = JNIHandles::resolve_non_null(jthread);
    if (java_lang_Thread::thread(thread_oop) != nullptr) {
      // Try to capture the live JavaThread in a ThreadsListHandle:
      ThreadsListHandle tlh; // Provides memory barrier
      JavaThread* thr;
      if ( (thr = java_lang_Thread::thread(thread_oop)) != nullptr) {
        assert(tlh.includes(thr), "must be");
        // The still live JavaThread is protected by the ThreadsListHandle
        // so is safe to access.
        Parker* p = thr->parker();
        HOTSPOT_THREAD_UNPARK((uintptr_t) p);
        p->unpark();
      }
    } // ThreadsListHandle is destroyed here.
  }
} UNSAFE_END
```

Reply:
It's also possible to create a special version of ThreadsListHandle which takes an Thread object in constructor.
Do the eetop check and then safe fetch the ThreadsList.
If you call include() with the same JavaThread* as the eetop was before creating the ThreadsList, you can return true directly, i.e. quick mode.

```
SpecialThreadsListHandle::SpecialThreadsListHandle(oop thread_oop) {
  _javaThread_from_eetop = java_lang_Thread::thread(thread_oop);
  _threadList = ...
}

bool includes(JavaThread* jt) {
  if (jt == _javaThread_from_eetop && jt != nullptr) return true;
  return _threadsList->includes(jt);
}
```

Reply:
Yes, we could create a special version of ThreadsListHandle that would allow us
to encapsulate this special purpose logic in threadsSMR source files where it
belongs. Of course, because I can't stand the 'eetop' name, I'll use some other
naming here. :-)

Reply:
Hmmm... it occurs me that this special version of ThreadsListHandle will need a
way to communicate with the call-site that created it and return the protected
`JavaThread*`. If it returns `nullptr`, then the `thread_oop` couldn't be
converted into a protected `JavaThread*`.

Perhaps `stlh.protected_java_thread()`...